### ISSUE TITLE
sbt testOnly command should fail/warn when no test found

### ISSUE THREAD

**Issue by pedrorijo91** [sbt/sbt#3188]
## problem

if you do a typo on the `sbt testOnly` command you will hardly notice the test you were expecting didn't run. 

Current output: 

```
(...)
[info] Run completed in 46 milliseconds.
[info] Total number of tests run: 0
[info] Suites: completed 0, aborted 0
[info] Tests: succeeded 0, failed 0, canceled 0, ignored 0, pending 0
[info] No tests were executed.
[info] No tests to run for test:testOnly
[success] Total time: 69 s, completed May 11, 2017 10:36:33 AM
```
notice the last line, which says success. If you do a quick read, you probably won't notice. At least myself, I just look for the green or red line at the bottom to see if everything it's ok.

## expectation

Why not fail? Or at least something more noticeable?

## notes

Available to contribute, just need some guidelines

**Comment by dwijnand** [comment]
I believe the code to look at is the value of `testFilter in testOnly`: https://github.com/sbt/sbt/blob/0b2081bc3cd3bb8677f042c2d7346175175f8ec7/main/src/main/scala/sbt/Defaults.scala#L584

Which leads to `selectedFilter`: https://github.com/sbt/sbt/blob/0b2081bc3cd3bb8677f042c2d7346175175f8ec7/main/src/main/scala/sbt/Defaults.scala#L873-L887

The only tricky thing I see is that `selectedFilter` is also used by `testQuick`, so changing it might or might not be correct.

**Comment by pedrorijo91** [comment]
Thanks for the guideline. going to find some time to have a look at the code then.

So I suppose you agree that a better output could exist? Would you rather a fail on for a better warning?

**Comment by jvican** [comment]
I would fail :smile:.

**Comment by eed3si9n** [comment]
If the root project aggregated subprojects `projA`, `projB`, and `projC`, and you type in `testOnly a.foo`, would you fail the whole operation because `a.foo` doesn't exist on `projB`?

**Comment by jvican** [comment]
Yes, if that's the case in the CI it means that the developer really wanted to run that test. Otherwise it wouldn't have typed the `testOnly a.foo`. The problematic with not failing is that this error could be missed for days, if not months. A warning is not enough -- no one looks at the logs when CI is green.

**Comment by jvican** [comment]
@pedrorijo91 Do you want us to assignt this task to you? :smile:

**Comment by pedrorijo91** [comment]
I would love to contribute. You can assign this to me. I hadn't a lot of free time to dig into sbt recently, but I think better days are coming :) (I already cloned and ran the tests, it's a good start 😛)

**Comment by eed3si9n** [comment]
> Yes, if that's the case in the CI it means that the developer really wanted to run that test.

I disagree with the expectation of this issue. I regularly run `testOnly` on the root project that aggregates multiple subproject (e.g. projA, projB, and projC) fully aware that the test only exists in one of the many subprojects.
If it raised warning or error, it would be a noise for that use case.

**Comment by jvican** [comment]
We can reconcile both scenarios: we make it fail when no test is run (across any subproject) and we make it succeed otherwise. This could be detected via `executionRoots`, I think.

**Comment by eed3si9n** [comment]
ok. If it's able to run a test out of `projB` on aggregated projA, projB, and projC, I'd be fine to fail on the case where the test example doesn't exist in any subprojects.

### note

Currently there's a separation of concerns between the task engine and the task. The task engine splits the work across aggregated subprojects and run them in parallel. Task implementation focuses on one subproject.

Likely there needs to be two step process where the first process finds the aggregated match of test examples. Then, if empty fail, otherwise let the test example run.

**Comment by xuwei-k** [comment]
Here is a solution. but there is a problem as @eed3si9n say

```scala
testResultLogger := TestResultLogger.Defaults.Main(
  printNoTests = TestResultLogger((_, _, _) => sys.error(""))
)
```

> If the root project aggregated subprojects projA, projB, and projC, and you type in testOnly a.foo, would you fail the whole operation because a.foo doesn't exist on projB?

**Comment by jvican** [comment]
> Currently there's a separation of concerns between the task engine and the task. The task engine splits the work across aggregated subprojects and run them in parallel. Task implementation focuses on one subproject.
  
Yes, but this doesn't mean that task implementations must not be aware of the internal state from sbt. In fact, the `update` implementation is an example of the solution I'm proposing, where `executionRoots` is used to know whether update should be forced or not.
  
@xuwei-k Yes, and it seems that's not the solution we're looking for :smile:.

**Comment by eed3si9n** [comment]
To implement this `testOnly` should become a command. Something that can/should be experimented as a plugin.

The command can mimic the aggregation semantics of act command and verify the test count.

**Comment by calm329** [comment]
@eed3si9n Please let me know if this issue is still available to work on

**Comment by eed3si9n** [comment]
@calm329 Yes, it's still available.
